### PR TITLE
Memory Pair Game

### PR DESCRIPTION
**Pull request by KikinovK**
### Memory Pair Game

[Demo](https://kikinovk.github.io/Memory_Pair_Game/) | 
[Code base](https://github.com/KikinovK/frontend-2021-homeworks/tree/memory-pair-game-task/submissions/kikinovk/Memory_Pair-game)

Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@KikinovK please double check the links you've provided

**Comment by KikinovK** [comment]
Checked link [Demo](https://kikinovk.github.io/Memory_Pair_Game/)

**Comment by OleksiyRudenko** [comment]
@KikinovK spotted a bug.
Steps to reproduce:
1. Click card A
2. Click card B
3. Click card C quick enough (while cards A & B are still either open or haven't closed back fully)

Expected result:
(a) If A === B then remove both
(b) If A === C or B === C then turn A & B back, keep C open

Actual result:
(a) if A === C or B === C then C and matching card (A or B) get removed and unmatching card (B or A) turns back

Illustrations:
Screenshot_2


Screenshot_3


Screenshot_4


Screenshot_5

**Comment by KikinovK** [comment]
The reason for the error was to clear the flag inside asynchrony (setTimeout).

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                      <img src="${src}" class="card__img--back">
+                    </div>`;
+  return card
+};
+
+const initGameBoard = () => {
+  const fragment = document.createDocumentFragment();
+
+  for (let i = 0; i < maxCard*2; i++) {
+    let card = creatCard(selectImage(cards, images));
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:59]:
```suggestion
    const card = createCard(selectImage(cards, images));
```

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  do {
+    random = Math.floor(Math.random() * Math.floor(maxRandom));
+  } while (cards
+              .filter(card => card.dataset.id === images[random].id)
+              .length === 2);
+  return images[random]
+};
+
+const creatCard = ({src, id}) => {
+  let card =  document.createElement('div');
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:42]: why let ?

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  card.classList.add('card');
+  card.dataset.id = id;
+  card.innerHTML = `<div class="card__front">
+                      <img src="./img/face.png" class="card__img--front" >
+                    </div>
+                    <div class="card__back">
+                      <img src="${src}" class="card__img--back">
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:50]: alt

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    card.classList.add('card--disappear')
+    setTimeout(() => {
+      card.classList.add('card--disable')
+    }, delay);
+  },  delay*2);
+};
+
+const eventCard = () => {
+  cards.forEach(card => {
+    card.addEventListener('click', () => {
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:91]: use event delegation

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      openedCard(card);
+      if (!openCard) {
+       openCard = card;
+       card.open = true;
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:95]: are you sure that your card have `open` property ?

**Review by babayK0 — CHANGES_REQUESTED** [review]
Good job !

File: submissions/kikinovk/Memory_Pair-game/js/main.js (hunk 2 above)
Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:42]: You are right, `card` is an immutable object link. Better than `const`

File: submissions/kikinovk/Memory_Pair-game/js/main.js (hunk 3 above)
Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:50]: What did you mean?

File: submissions/kikinovk/Memory_Pair-game/js/main.js (hunk 5 above)
Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: The `open` property is declared for the first time in this part of the code. I can't do that?

Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: Is it better to declare the `open` property in the `creatCard` function?

File: submissions/kikinovk/Memory_Pair-game/js/main.js (hunk 2 above)
Comment (babayK0) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:42]: always try to use `const`. Use `let` only for variables which should change

File: submissions/kikinovk/Memory_Pair-game/js/main.js (hunk 3 above)
Comment (babayK0) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:50]: [Html alt attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attributes)

File: submissions/kikinovk/Memory_Pair-game/js/main.js (hunk 5 above)
Comment (babayK0) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: what stored in `card` variable ?

Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: in `card` variable  the card itself from the `Cards` array
image

Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: image

Comment (babayK0) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: It is better to not mix render data (html) with properties/state and so on.

Comment (babayK0) [reply submissions/kikinovk/Memory_Pair-game/js/main.js:95]: [Separation_of_concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)

Comment (KikinovK) [reply submissions/kikinovk/Memory_Pair-game/js/main.js]: can I use `cards.classList.contains('.card--open') === true` instead of `card.open === true`?

Comment (babayK0) [reply submissions/kikinovk/Memory_Pair-game/js/main.js]: yep

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+const disappearCard = (card) => {
+  setTimeout(() => {
+    card.classList.add('card--disappear')
+    setTimeout(() => {
+      card.classList.add('card--disable')
+    }, delay);
+  },  delay*2);
+};
+
+const eventCard = () => {
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:95]: function names must be verbs

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    cards.push(card);
+  };
+  gameBoard.appendChild(fragment);
+};
+
+const openedCard = (card) => {
+  card.classList.add('card--open');
+};
+
+const closedCard = (card) => {
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:77]: verb

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  for (let i = 0; i < maxCard*2; i++) {
+    const card = creatCard(selectImage(cards, images));
+
+    fragment.appendChild(card);
+    cards.push(card);
+  };
+  gameBoard.appendChild(fragment);
+};
+
+const openedCard = (card) => {
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:73]: verb

File: submissions/kikinovk/Memory_Pair-game/js/main.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                      <img src="${src}" class="card__img--back" alt="${alt}">
+                    </div>`;
+  return card
+};
+
+const initGameBoard = () => {
+  const fragment = document.createDocumentFragment();
+
+  for (let i = 0; i < maxCard*2; i++) {
+    const card = creatCard(selectImage(cards, images));
```
Comment (babayK0) [inline submissions/kikinovk/Memory_Pair-game/js/main.js:65]:
```suggestion
    const card = createCard(selectImage(cards, images));
```

**Review by babayK0 — CHANGES_REQUESTED** [review]

**Review by babayK0 — APPROVED** [review]
:+1:

**Pull request merged**